File: test/vectors/verify-tool-integrity.mjs
#!/usr/bin/env node
/**
 * Cross-language conformance verifier for tool-definition canonicalization.
 * Verifies: SHA-256(JCS(toolDefinition)) matches expected digests.
 */
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import canonicalize from "canonicalize";

const __dirname = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(
  readFileSync(join(__dirname, "tool-definition-canonicalization.json"), "utf-8"),
);

function computeDigest(tool) {
  const canonical = canonicalize(tool);
  if (canonical === undefined) throw new Error("canonicalize returned undefined");
  return createHash("sha256").update(canonical).digest("hex");
}

let passed = 0;
let failed = 0;
let skipped = 0;

console.log("=== Positive vectors (expected digest must match) ===\n");
for (const vec of vectors.positive) {
  try {
    const actual = computeDigest(vec.tool);
    if (actual === vec.expectedDigest) {
      console.log(`  ✓ ${vec.id}: ${vec.description}`);
      passed++;
    } else {
      console.log(`  ✗ ${vec.id}: ${vec.description}`);
      console.log(`    expected: ${vec.expectedDigest}`);
      console.log(`    actual:   ${actual}`);
      failed++;
    }
  } catch (err) {
    console.log(`  ✗ ${vec.id}: ${vec.description} — ERROR: ${err.message}`);
    failed++;
  }
}

console.log("\n=== Negative vectors ===\n");
for (const vec of vectors.negative) {
  if (vec.type === "same_digest") {
    try {
      const dA = computeDigest(vec.toolA);
      const dB = computeDigest(vec.toolB);
      if (dA === dB) {
        console.log(`  ✓ ${vec.id}: ${vec.description}`);
        if (vec.expectedDigest && dA !== vec.expectedDigest) {
          console.log(`    WARNING: digests match each other but differ from expected`);
          console.log(`    expected: ${vec.expectedDigest}, actual: ${dA}`);
        }
        passed++;
      } else {
        console.log(`  ✗ ${vec.id}: ${vec.description}`);
        console.log(`    digestA: ${dA}`);
        console.log(`    digestB: ${dB}`);
        console.log(`    Expected same digest but got different`);
        failed++;
      }
    } catch (err) {
      console.log(`  ✗ ${vec.id}: ERROR: ${err.message}`);
      failed++;
    }
  } else if (vec.type === "different_digest") {
    try {
      const dA = computeDigest(vec.toolA);
      const dB = computeDigest(vec.toolB);
      if (dA !== dB) {
        console.log(`  ✓ ${vec.id}: ${vec.description}`);
        if (vec.digestA && dA !== vec.digestA) {
          console.log(`    WARNING: digestA mismatch: expected ${vec.digestA}, got ${dA}`);
        }
        if (vec.digestB && dB !== vec.digestB) {
          console.log(`    WARNING: digestB mismatch: expected ${vec.digestB}, got ${dB}`);
        }
        passed++;
      } else {
        console.log(`  ✗ ${vec.id}: ${vec.description}`);
        console.log(`    Both produced: ${dA}`);
        console.log(`    Expected different digests`);
        failed++;
      }
    } catch (err) {
      console.log(`  ✗ ${vec.id}: ERROR: ${err.message}`);
      failed++;
    }
  } else if (vec.type === "must_reject") {
    if (vec.tool) {
      try {
        computeDigest(vec.tool);
        // If we get here, check if the tool contains surrogates that should reject
        const toolStr = JSON.stringify(vec.tool);
        const hasSurrogate = /[\uD800-\uDFFF]/.test(toolStr);
        if (hasSurrogate) {
          console.log(`  ✗ ${vec.id}: ${vec.description} — should have rejected but produced a digest`);
          failed++;
        } else {
          // Value was normalized by JSON parse (e.g., >2^53 rounded) — can't test from JSON
          console.log(`  ~ ${vec.id}: ${vec.description} (value normalized at parse, skipped)`);
          skipped++;
        }
      } catch (err) {
        console.log(`  ✓ ${vec.id}: ${vec.description} (correctly rejected: ${err.message.slice(0, 60)})`);
        passed++;
      }
    } else {
      console.log(`  ~ ${vec.id}: ${vec.description} (no tool provided, skipped)`);
      skipped++;
    }
  }
}

console.log("\n=== Drift vectors (before/after digest comparison) ===\n");
for (const vec of vectors.drift) {
  try {
    const beforeD = computeDigest(vec.before);
    const afterD = computeDigest(vec.after);
    const driftDetected = beforeD !== afterD;

    if (driftDetected === vec.driftExpected) {
      console.log(`  ✓ ${vec.id}: ${vec.description}`);
      if (vec.beforeDigest && beforeD !== vec.beforeDigest) {
        console.log(`    WARNING: beforeDigest mismatch`);
      }
      if (vec.afterDigest && afterD !== vec.afterDigest) {
        console.log(`    WARNING: afterDigest mismatch`);
      }
      passed++;
    } else {
      console.log(`  ✗ ${vec.id}: ${vec.description}`);
      console.log(`    before: ${beforeD}`);
      console.log(`    after:  ${afterD}`);
      console.log(`    drift detected: ${driftDetected}, expected: ${vec.driftExpected}`);
      failed++;
    }
  } catch (err) {
    console.log(`  ✗ ${vec.id}: ERROR: ${err.message}`);
    failed++;
  }
}

console.log(`\n=== Results ===`);
console.log(`  Passed:  ${passed}`);
console.log(`  Failed:  ${failed}`);
console.log(`  Skipped: ${skipped}`);
console.log(`  Total:   ${passed + failed + skipped}`);

if (failed > 0) {
  process.exit(1);
}

File: test/vectors/verify-tool-integrity.py
#!/usr/bin/env python3
"""
Cross-language conformance verifier for tool-definition canonicalization.
Verifies: SHA-256(JCS(toolDefinition)) matches expected digests.

Requires: pip install rfc8785
(NOT canonicaljson — that's Matrix canonical JSON, rejects floats)
"""
import hashlib
import json
import os
import sys

try:
    import rfc8785
except ImportError:
    print("ERROR: 'rfc8785' package required. Install with: pip install rfc8785")
    print("       Do NOT use 'canonicaljson' — it's Matrix canonical JSON, not RFC 8785.")
    sys.exit(2)


def compute_digest(tool: dict) -> str:
    canonical = rfc8785.dumps(tool)
    return hashlib.sha256(canonical).hexdigest()


def main():
    vectors_path = os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "tool-definition-canonicalization.json",
    )
    with open(vectors_path, "r", encoding="utf-8") as f:
        vectors = json.load(f)

    passed = 0
    failed = 0
    skipped = 0

    print("=== Positive vectors (expected digest must match) ===\n")
    for vec in vectors["positive"]:
        try:
            actual = compute_digest(vec["tool"])
            if actual == vec["expectedDigest"]:
                print(f"  ✓ {vec['id']}: {vec['description']}")
                passed += 1
            else:
                print(f"  ✗ {vec['id']}: {vec['description']}")
                print(f"    expected: {vec['expectedDigest']}")
                print(f"    actual:   {actual}")
                failed += 1
        except Exception as e:
            print(f"  ✗ {vec['id']}: {vec['description']} — ERROR: {e}")
            failed += 1

    print("\n=== Negative vectors ===\n")
    for vec in vectors["negative"]:
        vtype = vec["type"]
        if vtype == "same_digest":
            try:
                d_a = compute_digest(vec["toolA"])
                d_b = compute_digest(vec["toolB"])
                if d_a == d_b:
                    print(f"  ✓ {vec['id']}: {vec['description']}")
                    if "expectedDigest" in vec and d_a != vec["expectedDigest"]:
                        print(f"    WARNING: digests match but differ from expected")
                    passed += 1
                else:
                    print(f"  ✗ {vec['id']}: {vec['description']}")
                    print(f"    digestA: {d_a}")
                    print(f"    digestB: {d_b}")
                    print(f"    Expected same digest but got different")
                    failed += 1
            except Exception as e:
                print(f"  ✗ {vec['id']}: ERROR: {e}")
                failed += 1
        elif vtype == "different_digest":
            try:
                d_a = compute_digest(vec["toolA"])
                d_b = compute_digest(vec["toolB"])
                if d_a != d_b:
                    print(f"  ✓ {vec['id']}: {vec['description']}")
                    if "digestA" in vec and d_a != vec["digestA"]:
                        print(f"    WARNING: digestA mismatch: expected {vec['digestA']}, got {d_a}")
                    if "digestB" in vec and d_b != vec["digestB"]:
                        print(f"    WARNING: digestB mismatch: expected {vec['digestB']}, got {d_b}")
                    passed += 1
                else:
                    print(f"  ✗ {vec['id']}: {vec['description']}")
                    print(f"    Both produced: {d_a}")
                    print(f"    Expected different digests")
                    failed += 1
            except Exception as e:
                print(f"  ✗ {vec['id']}: ERROR: {e}")
                failed += 1
        elif vtype == "must_reject":
            print(f"  ~ {vec['id']}: {vec['description']} (validation-level, skipped in JSON verifier)")
            skipped += 1

    print("\n=== Drift vectors (before/after digest comparison) ===\n")
    for vec in vectors["drift"]:
        try:
            before_d = compute_digest(vec["before"])
            after_d = compute_digest(vec["after"])
            drift_detected = before_d != after_d

            if drift_detected == vec["driftExpected"]:
                print(f"  ✓ {vec['id']}: {vec['description']}")
                if "beforeDigest" in vec and before_d != vec["beforeDigest"]:
                    print(f"    WARNING: beforeDigest mismatch")
                if "afterDigest" in vec and after_d != vec["afterDigest"]:
                    print(f"    WARNING: afterDigest mismatch")
                passed += 1
            else:
                print(f"  ✗ {vec['id']}: {vec['description']}")
                print(f"    before: {before_d}")
                print(f"    after:  {after_d}")
                print(f"    drift detected: {drift_detected}, expected: {vec['driftExpected']}")
                failed += 1
        except Exception as e:
            print(f"  ✗ {vec['id']}: ERROR: {e}")
            failed += 1

    print(f"\n=== Results ===")
    print(f"  Passed:  {passed}")
    print(f"  Failed:  {failed}")
    print(f"  Skipped: {skipped}")
    print(f"  Total:   {passed + failed + skipped}")

    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    main()
